File: BACKEND/admin/services/staffService.js
const adminRepository = require("../models/adminRepository");
const { createServiceError } = require("./errors");

const allowedStatuses = new Set(["active", "inactive", "on_leave"]);

async function listStaff(filters = {}) {
  return adminRepository.listStaff(filters);
}

async function getStaffMemberById(id) {
  const member = await adminRepository.getStaffMemberById(id);
  if (!member) {
    throw createServiceError(404, "Staff member not found.");
  }
  return member;
}

function validateStaffPayload(payload) {
  const errors = {};

  if (payload.expertise !== undefined && String(payload.expertise).trim().length === 0) {
    errors.expertise = "Chuyen mon khong duoc rong.";
  }
  if (payload.yearsOfExperience !== undefined) {
    const years = Number(payload.yearsOfExperience);
    if (!Number.isFinite(years) || years < 0 || years > 80) {
      errors.yearsOfExperience = "Nam kinh nghiem phai la so tu 0 den 80.";
    }
  }
  if (payload.status !== undefined && !allowedStatuses.has(payload.status)) {
    errors.status = "Status must be active, inactive, or on_leave.";
  }

  if (Object.keys(errors).length) {
    throw createServiceError(422, "Staff payload is invalid.", errors);
  }
}

async function updateStaffMember(id, payload) {
  validateStaffPayload(payload);
  await getStaffMemberById(id); // dam bao member ton tai
  const updated = await adminRepository.updateStaffMember(id, payload);
  if (!updated) {
    throw createServiceError(404, "Staff member not found.");
  }
  return updated;
}

async function setStaffAvailability(id, status) {
  if (!allowedStatuses.has(status)) {
    throw createServiceError(422, "Status must be active, inactive, or on_leave.");
  }
  return updateStaffMember(id, { status });
}

module.exports = {
  listStaff,
  getStaffMemberById,
  updateStaffMember,
  setStaffAvailability,
};

File: BACKEND/admin/Services/StaffService.cs
using MyPuppy.Admin.Models;
using MyPuppy.Admin.Repositories;

namespace MyPuppy.Admin.Services;

public class StaffService
{
    private static readonly HashSet<string> AllowedStatuses = new(StringComparer.OrdinalIgnoreCase)
    {
        "active", "inactive", "on_leave"
    };

    private readonly AdminRepository _repository;

    public StaffService(AdminRepository repository)
    {
        _repository = repository;
    }

    public Task<IEnumerable<StaffDto>> ListStaffAsync(string? status, string? search)
        => _repository.ListStaffAsync(status, search);

    public async Task<StaffDto> GetStaffByIdAsync(string id)
    {
        var staff = await _repository.GetStaffByIdAsync(id);
        if (staff == null) throw new ServiceException(404, "Staff member not found.");
        return staff;
    }

    public async Task<StaffDto> UpdateStaffAsync(string id, UpdateStaffRequest payload)
    {
        ValidatePayload(payload);
        await GetStaffByIdAsync(id);
        var updated = await _repository.UpdateStaffAsync(id, payload);
        return updated ?? throw new ServiceException(404, "Staff member not found.");
    }

    public Task<StaffDto> SetAvailabilityAsync(string id, string status)
    {
        if (string.IsNullOrWhiteSpace(status) || !AllowedStatuses.Contains(status))
            throw new ServiceException(422, "Status must be active, inactive, or on_leave.");

        return UpdateStaffAsync(id, new UpdateStaffRequest(null, null, status));
    }

    private static void ValidatePayload(UpdateStaffRequest payload)
    {
        var errors = new Dictionary<string, string>();

        if (payload.Expertise != null && payload.Expertise.Trim().Length == 0)
            errors["expertise"] = "Chuyen mon khong duoc rong.";

        if (payload.YearsOfExperience.HasValue)
        {
            var years = payload.YearsOfExperience.Value;
            if (years < 0 || years > 80)
                errors["yearsOfExperience"] = "Nam kinh nghiem phai la so tu 0 den 80.";
        }

        if (payload.Status != null && !AllowedStatuses.Contains(payload.Status))
            errors["status"] = "Status must be active, inactive, or on_leave.";

        if (errors.Count > 0)
            throw new ServiceException(422, "Staff payload is invalid.", errors);
    }
}
